### ISSUE TITLE
Cannot compile optee_test as a standalone build

### ISSUE THREAD

**Issue by vikaspot** [OP-TEE/optee_test#334]
Hi all,
I am facing issue while building for optee_test  
Error:
  CC      optee_test/host/xtest/install_ta.c
  LD      optee_test/out/xtest/xtest
../openssl/lib/aarch64/libcrypto.a: error adding symbols: File format not recognized

Can anyone please help ?
What is the openssl version is taken  and crosscompiled and kept under /openssl/lib/aarch64/ 

How's the optee roadmap is decided like... Optee_os/test/. version . With respect to openssl Version  mapping  ?
../openssl/lib/aarch64/libcrypto.a: which version of Opensssl is this binary kept ?

Thanks 
Vikas

**Comment by jbech-linaro** [comment]
Are you trying to compile `optee_test` standalone or using [build.git](https://optee.readthedocs.io/building/gits/build.html#build)? If it's the former, then yes, it's a bit challenging due to various dependencies (OpenSSL one of them). If it's the latter, then it should work "out of the box".

Building gits purely standalone is something I personally would like to see getting improved in OP-TEE (and we are aware of the challenges).

**Comment by vikaspot** [comment]
I am trying to build stand alone optee_test.

**Comment by vikaspot** [comment]
Can you please tell me which openssl version is compatible for optee 3.3.0 ?
is there any mapping info is available

**Comment by jbech-linaro** [comment]
I'm not aware of us specifying a certain version. We are using the one that comes with the Buildroot package (enabled [here](https://github.com/OP-TEE/build/blob/master/common.mk#L236-L237) in build.git). @jenswi-linaro am I right or wrong here?

> Can you please tell me which openssl version is compatible for optee 3.3.0 ?

Seems to pick `1.0.2p` from Buildroot:

```
-- Found OpenSSL: /media/jbech/SSHD_LINUX/devel/optee_projects/qemu/out-br/host/arm-buildroot-linux-gnueabihf/sysroot/usr/lib/libssl.so;/media/jbech/SSHD_LINUX/devel/optee_projects/qemu/out-br/host/arm-buildroot-linux-gnueabihf/sysroot/usr/lib/libcrypto.so (found version "1.0.2p")
```

Having that said, I gave it a try to build optee_test "stand-alone" using Cmake in hope of being able to give you some guidance. Unfortunately I failed on this first try (compiling fails in pointing to the h-files in optee_client). I could probably sort it out if I give it some time (which I don't have for the moment). I believe we have to put this on the to-do list of things to improve/sort out in the (near) future.

**Comment by jenswi-linaro** [comment]
@jbech-linaro you're right, we use whatever version of OpenSSL that buildroot provides by default.

**Comment by jbech-linaro** [comment]
@vikaspot I'm not sure if you intend to try to come up with some solution for this on your own. If you do, then feel free to also update the documentation page related to this (see [optee_test in our docs](https://optee.readthedocs.io/building/gits/optee_test.html#build-instructions)).

**Comment by jbech-linaro** [comment]
I took the liberty to rename the name of the issue and assign to the label "enhancement". I.e., mark it as something that has to be fixed.

**Comment by vikaspot** [comment]
Thanks for guidance.
As per my understanding .
We should have have mapping for version of Optee_os /Optee_client/Optee_driver.... etc all  .. with all open sources components Version which can be assumed as stable ... image in combination.

-- Ok .. I'll give a try to build for openssl cross compiling for my SOC and get it linked with optee_test.

Once every thing got done i'll attempt to update in optee/Docs
.
  OpenSSL_1_0_2p  is  released on Aug 14, 2018
and optee_test 3.3.0 is release on Oct 12, 2018
so  assuming i should used openssl 1.0.2P version and cross compile for my target and place the library @  /openssl/lib/aarch64/ location then it should be able to work properly ...
@jbech-linaro  please correct me if i am wrong !! hope Version compatibility will be there :)

**Comment by jbech-linaro** [comment]
> We should have have mapping for version of Optee_os /Optee_client/Optee_driver.... etc all .. with all open sources components Version which can be assumed as stable ... image in combination.

We have that already, that is the stable builds (see [this](https://optee.readthedocs.io/building/gits/build.html#stable-releases)). So if you open any of the [xml-files in a branches corresponding to a release](https://github.com/OP-TEE/manifest/branches/all), then you will see which versions that are compatible with each other. Example:
- Branch: https://github.com/OP-TEE/manifest/tree/3.4.0
- QEMU: https://github.com/OP-TEE/manifest/blob/3.4.0/default.xml <-- There you can see labels or a fixed commit if there were no suitable label(s) to use when we did the release. For OpenSSL, that means to figure out the mapping between Buildroot and what OpenSSL version Buildroot pointed at that particular git label.
I.e., the takeaway is that a stable xml only contains fixed labels/commit, i.e., never changes.

OpenSSL is a bit unique here, since I believe that it's the only dependency we have that is pulled in as a sub-component to a git we are dependent on. All other gits are pulled as separate gits as stated in the xml-files.

In short it's pretty messy for the moment when it comes to doing "standalone" builds.

**Comment by jbech-linaro** [comment]
Maybe I have something that is good enough for you to start with (half pseudo code)
```
$ git clone ... optee_client
$ git clone ... optee_os
$ git clone ... optee_test

$ cd optee_client
$ make INCLUDEDIR=/include LIBDIR=/lib

$ cd optee_os
$ make optee-os CFG_TA_MBEDTLS=n CFG_TA_MBEDTLS_MPI=n <other-flags-that-you-are-using>

$ cd optee_test
$ make \
        CROSS_COMPILE_HOST=arm-linux-gnueabihf- \
        CROSS_COMPILE_TA=arm-linux-gnueabihf- \
        TA_DEV_KIT_DIR=<path-to-os>/optee_os/out/arm/export-ta_arm32 \
        TEEC_EXPORT=<path-to-client>/optee_client/out/export/usr \
        CFG_TA_MBEDTLS=n \
        CFG_TA_MBEDTLS_MPI=n
```
With that I could successfully compile `optee_test` using the old plain Makefiles. Tests that are using OpenSSL are excluded when doing like this.

**Comment by jbech-linaro** [comment]
I've put together a toy example here, that does what I did above. I will use that as a reminder to myself for now:
https://github.com/jbech-linaro/optee_standalone

**Comment by vikaspot** [comment]
@jbech-linaro 
Thanks for creating a different  branch for standalone compilation .. i'll try today .. 

Is it possible to share the log of openssl compilation within  buildroot for aarch64.

I need to see ./Configure .... - options for openssl while building ..

**Comment by jbech-linaro** [comment]
I've done some progess with CMake also. However, it's impossible to do everything in CMake at the moment, since there are dependencies to files generated by OP-TEE OS. The files are direct included by optee_test CMake files. I've put together a setup where I do as little as possible using GNU Make and do as much as possible using CMake. I've put that in a separate branch which you can find here: https://github.com/jbech-linaro/optee_standalone/tree/cmake_optee .

**Comment by vikaspot** [comment]
I am able to cross compile openssl for my soc armv8 and cortex-a7 .
I believe there is some issue with my build environment .it seems while building whole bunch of optee components it for aarch64 it takes in bw a 32 cross compiler .. so compatibility issue is faced by me
